### ISSUE TITLE
Quantize weights causes accuracy to plunge when run in mobile but not in computers?

### ISSUE THREAD

**Issue by kwotsin** [tensorflow/tensorflow#11182]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: Source
- **TensorFlow version (use command below)**: 1.2.1
- **Python version**: 
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:
```
/home/kwotsin/tensorflow/bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
--in_graph=./frozen_inception_resnet_v2_for_mobile.pb \
--out_graph=./quantized_inception_resnet_v2_for_mobile_NEW.pb \
--inputs='Placeholder_only' \
--outputs='InceptionResnetV2/Logits/Predictions' \
--transforms='
  add_default_attributes
  strip_unused_nodes(type=float, shape="1,299,299,3")
  remove_nodes(op=Identity, op=CheckNumerics)
  fold_constants(ignore_errors=true)
  fold_batch_norms
  fold_old_batch_norms
  quantize_weights
  strip_unused_nodes
  sort_by_execution_order'
```

### Describe the problem
Using the above quantization method, the quantization tool works so well that there is hardly a noticeable difference in accuracy drop (less than 0.5%) for a model like inception v3, with a 1/4 size reduction and slightly faster speed. However, when using the exact same files to be run on mobile, the performance gets so poor that there's more than 70-80% accuracy decrease. I'm unsure whether the issue lies with the quantization not getting optimized on mobile architectures (ARM instead of the usual desktop amd architecture), or whether there is a problem in the operations for the tensorflow mobile library.

Note that `quantize_nodes` is totally unusable. When used to quantize the model, the model size increases a little and then causes the app to crash instantly. The error log produced when using quantize_nodes is this:

```
07-01 00:07:01.760 28272-28357/com.mindorks.tensorflowexample E/art: No implementation found for long org.tensorflow.contrib.android.RunStats.allocate() (tried Java_org_tensorflow_contrib_android_RunStats_allocate and Java_org_tensorflow_contrib_android_RunStats_allocate__)
07-01 00:07:03.508 28272-28357/com.mindorks.tensorflowexample A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0x0 in tid 28357 (pool-1-thread-1)
```

Also, for almost every model I ran, the following error appeared:
`No implementation found for long org.tensorflow.contrib.android.RunStats.allocate()`
What does this mean and how could I resolve it?

FYI: Not sure if it makes a difference, but when I built my lib_tensorflow_inference.so file and the JAR file for using the TF library on mobile, the tensorflow version was cloned from the master branch and not git checked out. Would this make a difference?

Further weird phenomenon:

Although I built my TF from source and bazel built the graph transform tool, the following warnings still appear:

```
2017-07-01 00:05:19.612228: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2017-07-01 00:05:19.612262: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2017-07-01 00:05:19.612278: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2017-07-01 00:05:19.612282: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2017-07-01 00:05:19.612290: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
```

Thank you.

**Comment by ali01** [comment]
Could you please provide us with all available information about the mobile architecture/platform you're targeting?

**Comment by kwotsin** [comment]
Oh yes indeed I'm working on an android LG G3 phone that's running ARM architecture I believe.

**Comment by kwotsin** [comment]
I might have just solved this problem by looking at what #8897 did. Basically, I had to include the argument `--copt="-DTENSORFLOW_DISABLE_META" ` when building the `libtensorflow_inference.so` file. This reduced the library size for android and made the performance almost exactly as expected when run on a desktop.

Here is the full command to build libtensorflow_inference.so:
```
bazel build -c opt --copt="-DTENSORFLOW_DISABLE_META"  //tensorflow/contrib/android:libtensorflow_inference.so    --crosstool_top=//external:android/crosstool    --host_crosstool_top=@bazel_tools//tools/cpp:toolchain    --cpu=armeabi-v7a
```

and then the JAR file subsequently:

```
bazel build //tensorflow/contrib/android:android_tensorflow_inference_java
```

Further, I think there are two problematic transformations: Remove Nodes and Quantize Nodes. Using either one causes the accuracy to be reduced or inference timing to increase by around 5x.
The remaining 8 transformations are what I found as optimal:

```
/home/kwotsin/tensorflow/bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
--in_graph=./frozen_inception_resnet_v2_for_mobile.pb \
--out_graph=./quantized_inception_resnet_v2_for_mobile.pb \
--inputs='Placeholder_only' \
--outputs='InceptionResnetV2/Logits/Predictions' \
--transforms='
  add_default_attributes
  strip_unused_nodes(type=float, shape="1,299,299,3")
  fold_constants(ignore_errors=true)
  fold_batch_norms
  fold_old_batch_norms
  quantize_weights
  strip_unused_nodes
  sort_by_execution_order'
```

**Comment by Androbin** [comment]
@kwotsin 
It seems that the remove transformation also drops output nodes #8181. 
And sometimes removing them damages certain layers (e.g. batch norms) which happen to use some Identity ops which cannot be removed unconditionally.

**Comment by kwotsin** [comment]
Thanks for the explanation! Hopefully there's a fix coming soon

**Comment by cy89** [comment]
@kwotsin is the bug resolved for you? The other threads that @androbin references (#8181 and #10579) suggest that there are still some issues being worked out.

**Comment by kwotsin** [comment]
@androbin not sure if it matters, but I am using gcc and not clang when I configured my TF version, + python 2.7; also, I git checked out r1.2 instead of the master branch for the build. 

@cy89 yes I think the bug is resolved for me so far. I am seeing rather good results coming from the quantization when I remove those two problematic transformation - `remove_nodes` and `quantize_nodes`

**Comment by Androbin** [comment]
@kwotsin 
I am currently using clang, so will try with gcc now, thanks for the hint!
The version of python (3.5.2 in my case) should not matter?
Last pulled git commit is 454f704b094dee26a31d073c02f488cc8dce5944 from master.

**Comment by michaelisard** [comment]
@Androbin please let us know how this goes: I will close the issue if it's solved for you.

**Comment by Androbin** [comment]
Rebuilt everything from latest master, last error section changed slightly:
I found `zero_copy_stream` to be replaced by `coded_stream` but nothing else changed.

**Comment by martinwicke** [comment]
Pete, who is best to look at graph_transform issues, other than you?

**Comment by petewarden** [comment]
In this case, I believe the problem is related to metagemm, since @kwotsin references `--copt="-DTENSORFLOW_DISABLE_META"` as a fix for a similar problem, and the symptoms make sense. Passing to Maciek to investigate that side.

**Comment by kwotsin** [comment]
@Androbin It is quite strange the commands didn't work for you. Perhaps it is a system issue? Please let me know if you require information from my system build; I'll be glad to help out.

**Comment by Androbin** [comment]
@petewarden 
That sounds plausible, could you explain, what `metagemm` should do and what it is that is not working as intended? And are there any shortcomings of not having it running?

@kwotsin 
That is improbable, I have a pretty clean setup, not really any other major complaints. TensorFlow and its other tools worked just well till now. But quantization, selective registration and a few other tools have been flawky (for me and others) for quite some time now.

**Comment by maciekcc** [comment]
Hi there!
Metagemm is a library of high performance Arm32/64 kernels for quantized matrix multiplication, convolutions and related primitives. If turned off reference implementations will be used for some ops so you could expect lower performance.

I believe I might have let an old bug slip past. Metagemm right now has an aggressive code path for matrix multiplication and convolution with depth under 2048. And there is no test to actually check if k <= 2048.  If k > 2048, then depending on actual data you have some aggregators might overflow.

I should have a patch tomorrow, so you can see if that is actually the case.

**Comment by maciekcc** [comment]
Ok the patch should be in the internal Google repository soon and in github not long after.

**Comment by futurely** [comment]
https://github.com/tensorflow/tensorflow/commit/bdb2967a298236e24011405907cd19737386934e

**Comment by gxdalu-yaya** [comment]
@kwotsin hi，I use the DTENSORFLOW_DISABLE_META，the result is better, but the predict score is still diffrent with the python script's result.  I want to know , are your android result and python result is compeletly same?

**Comment by aselle** [comment]
Is this resolved?

**Comment by ljdang** [comment]
@aselle No, it is not resolved. Still need to add the argument --copt="-DTENSORFLOW_DISABLE_META" when building the "libtensorflow_inference.so" file. But this made the performance slower. When using the nightly-android builds, the result is different between mobile and PC.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue? Please update the label and/or status accordingly.

**Comment by jeffxtang** [comment]
Anyone having the low accuracy problem on iOS? I use `transform_graph` on a retrain model (based on Inception v3) and `label_image` works fine on the quantized model `transform_dog_retrained.pb` (same accuracy as with the pre-transformed model `dog_retrained.pb`), but the iOS app gives bad recognition results. The iOS app gives the same results as `label_image` when using the retrained (but before transformed) model `dog_retrained.pb`.

Commands to retrain, transform, and call label_image are as follows:

```
python tensorflow/examples/image_retraining/retrain.py   \
--model_dir=/tf_files/inception-v3   \
--output_graph=/tf_files/retrained_models/dog_retrained.pb   \
--output_labels=/tf_files/retrained_models/dog_retrained_labels.txt   \
--image_dir ~/Downloads/dog_images   \
--bottleneck_dir=/tf_files/dogs_bottleneck
```

```
bazel-bin/tensorflow/tools/graph_transforms/transform_graph  \
 --in_graph=/tf_files/retrained_models/dog_retrained.pb   \
--out_graph=/tf_files/retrained_models/transform_dog_retrained.pb   \
--inputs='Mul'   --outputs='final_result'   \
--transforms='
  add_default_attributes
  strip_unused_nodes(type=float, shape="1,299,299,3")
  fold_constants(ignore_errors=true)
  fold_batch_norms
  fold_old_batch_norms
  quantize_weights
  strip_unused_nodes
  sort_by_execution_order'
```
```
bazel-bin/tensorflow/examples/label_image/label_image \
--graph=/tf_files/retrained_models/transform_dog_retrained.pb \
--image=/tf_files/lab.jpg --input_layer=Mul --output_layer=final_result \
--labels=/tf_files/retrained_models/dog_retrained_labels.txt
```

Anyone has any idea how to fix the iOS issue? Btw, if I use the older `bazel build tensorflow/python/tools:strip_unused` then `python tensorflow/tools/quantization/quantize_graph.py`, the quantized model generated this way has good results (same accuracy as `label_image`) on iOS. But I hope to use the new `transform_graph` method instead of the old way.

What's the right way to use `transform_graph` so the generated model works on iOS with about the same accuracy as using`label_image`?

Thanks!

**Comment by aselle** [comment]
@petewarden, do we have someone who can take a look at this since maciek is no longer working on it.

**Comment by jeffxtang** [comment]
Can @petewarden or someone please take a look? Also, a related problem: I used one quantized MobileNet model to do the retraining:
```
python tensorflow/examples/image_retraining/retrain.py \
  --output_graph=/tf_files/dog_retrained_mobilenet10_224.pb \
  --output_labels=/tf_files/dog_retrained_labels_mobilenet.txt \
  --image_dir ~/Downloads/Images \
  --bottleneck_dir=/tf_files/dogs_bottleneck_mobilenet \
  --architecture mobilenet_1.0_224_quantized
```
and then using the label_image script to verify the model's accuracy, which looks pretty good:
```
bazel-bin/tensorflow/examples/label_image/label_image \
--graph=/tf_files/dog_retrained_mobilenet10_224.pb \
--image=/tmp/lab1.jpg \
--input_layer=input \
--output_layer=final_result \
--labels=/tf_files/dog_retrained_labels_mobilenet.txt \
--input_height=224 \
--input_width=224 \
--input_mean=128 \
--input_std=128
```
But when I use the retrained model in both iOS and Android, I have to set the threshold to 0.01 in order to see top recognition results, which look ok except the very low confidence values - for example: 
```
Predictions: 41 0.0261  n02099712 labrador retriever
60 0.024  n02106166 border collie
34 0.0227  n02099849 chesapeake bay retriever
31 0.0188  n02088238 basset
42 0.0174  n02091032 italian greyhound
``` 

The non-quantized version of a retrained MobileNet model (using `--architecture mobilenet_1.0_224` on retraining) shows good top results with high confidence values for the results in iOS and Android - for example:
```

Predictions: 5 0.328  n02091134 whippet
41 0.307  n02099712 labrador retriever
20 0.0988  n02088364 beagle
105 0.0815  n02087394 rhodesian ridgeback
32 0.0209  n02090379 redbone
```

I'm using TensorFlow 1.4 and the iOS `simple` example and the Android `TF Classify` example.

Any ideas on how to fix this and make retrained quantized MobileNet models work in iOS and Android? Thanks!

**Comment by petewarden** [comment]
We're now moving over to TF Lite as the recommended way to run quantized models, partly because it is less complex to integrate eight-bit with than classic TensorFlow. We have a demo of running the MobileNet image classification network on Android, and there should soon be full documentation on training it. I'm going to close this bug as not likely to be fixed, since our efforts will be focused on the TF Lite path.

**Comment by eaigner** [comment]
Yeah, well TF Lite is nowhere near where we need it to be to run our graphs. So what's the recommended solution until we can convert graphs to TF Lite? (trying to convert gives errors for unsupported ops).
